File: sdks/go/runtime_api.go
package memnox

import "context"

// Action verbs the Runtime API helpers check on the caller's behalf.
const (
	// ActionAccess is the verb CanAccess checks.
	ActionAccess = "resource.read"
	// ActionDeploy is the verb CanDeploy checks.
	ActionDeploy = "deploy.service"
	// ActionModify is the verb CanModify checks.
	ActionModify = "code.modify"
	// ActionDelete is the verb CanDelete checks.
	ActionDelete = "resource.delete"
)

// Verdict is a boolean answer plus the decision it came from.
type Verdict struct {
	// Allowed is true only when the runtime permitted the action outright.
	Allowed bool
	// Decision is the full verdict, including reason and matched policies.
	Decision Decision
}

// ShouldExecute answers whether one arbitrary action may run right now.
func (c *Client) ShouldExecute(ctx context.Context, request ActionRequest) (Verdict, error) {
	decision, err := c.Check(ctx, request)
	if err != nil {
		return Verdict{}, err
	}
	return Verdict{Allowed: decision.Allowed(), Decision: decision}, nil
}

// CanAccess answers whether the agent may read the given resource.
func (c *Client) CanAccess(ctx context.Context, resource, environment string) (Verdict, error) {
	return c.ShouldExecute(ctx, ActionRequest{Action: ActionAccess, Target: resource, Environment: environment})
}

// CanDeploy answers whether the agent may deploy the given service.
func (c *Client) CanDeploy(ctx context.Context, service, environment string) (Verdict, error) {
	return c.ShouldExecute(ctx, ActionRequest{Action: ActionDeploy, Target: service, Environment: environment})
}

// CanModify answers whether the agent may modify the given file or record.
func (c *Client) CanModify(ctx context.Context, target, environment string) (Verdict, error) {
	return c.ShouldExecute(ctx, ActionRequest{Action: ActionModify, Target: target, Environment: environment})
}

// CanDelete answers whether the agent may delete the given resource.
func (c *Client) CanDelete(ctx context.Context, target, environment string) (Verdict, error) {
	return c.ShouldExecute(ctx, ActionRequest{Action: ActionDelete, Target: target, Environment: environment})
}

File: packages/sdk/src/runtime-api.ts
import type { ActionRequest, RiskAssessment } from '@memnox/core';
import { DECISION_EFFECT } from '@memnox/core';
import type { MemnoxClient } from './client';

/** Boolean predicates over check(). Pending approval counts as "not permitted". */
export const RUNTIME_ACTION = {
  ACCESS: 'resource.access',
  DEPLOY: 'deploy.service',
  MODIFY: 'code.modify',
  DELETE: 'resource.delete',
} as const;

export interface ResourceQuery {
  target?: string;
  environment?: string;
  sessionId?: string;
}

function toRequest(action: string, query: ResourceQuery): ActionRequest {
  return {
    action,
    ...(query.target ? { target: query.target } : {}),
    ...(query.environment ? { environment: query.environment } : {}),
    ...(query.sessionId ? { sessionId: query.sessionId } : {}),
  };
}

export class RuntimeApi {
  constructor(private readonly client: MemnoxClient) {}

  /** True only when the action may proceed right now, unattended. */
  async shouldExecute(request: ActionRequest): Promise<boolean> {
    const decision = await this.client.check(request);
    return decision.effect === DECISION_EFFECT.ALLOW;
  }

  canAccess(query: ResourceQuery): Promise<boolean> {
    return this.shouldExecute(toRequest(RUNTIME_ACTION.ACCESS, query));
  }

  canDeploy(query: ResourceQuery): Promise<boolean> {
    return this.shouldExecute(toRequest(RUNTIME_ACTION.DEPLOY, query));
  }

  canModify(query: ResourceQuery): Promise<boolean> {
    return this.shouldExecute(toRequest(RUNTIME_ACTION.MODIFY, query));
  }

  canDelete(query: ResourceQuery): Promise<boolean> {
    return this.shouldExecute(toRequest(RUNTIME_ACTION.DELETE, query));
  }

  /** What would happen, without it happening — nothing is audited. */
  evaluateRisk(request: ActionRequest): Promise<RiskAssessment> {
    return this.client.evaluateRisk(request);
  }
}

export interface PolicyApplyResult {
  applied: boolean;
  version: string;
  policyCount: number;
  policyNames: string[];
}

export interface PolicyReloadResult {
  reloaded: boolean;
  version: string;
  /** A caller that just wrote a rule file checks this rather than assuming a reload. */
  sources?: string[];
}

export interface PolicySetView {
  version: string;
  policyCount: number;
  policyNames: string[];
  policies: unknown[];
}
